diff --git a/src/logger.cc b/src/logger.cc
index 6ddbea9..9a04b8a 100644
--- a/src/logger.cc
+++ b/src/logger.cc
@@ -85,11 +85,11 @@ std::string logger::err()
     char buf[2048];
 
 #if (_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600) && ! _GNU_SOURCE
-    if (strerror_r(errno, buf, sizeof(buf))
+    if (strerror_r_wrapper(strerror_r(errno, buf, sizeof(buf))), buf)
         return "Unknown error";
     return buf;
 #else
-    return strerror_r(errno, buf, sizeof(buf));
+    return strerror_r_wrapper(strerror_r(errno, buf, sizeof(buf)), buf);
 #endif
 }
 
@@ -228,4 +228,16 @@ bool logger::verbosity(const std::string& name)
     return false;
 }
 
+// XSI-compliant: int
+std::string logger::strerror_r_wrapper(int, char* s)
+{
+    return s;
+}
+
+// GNU-specific: char *
+std::string logger::strerror_r_wrapper(char* s, char*)
+{
+    return s;
+}
+
 NDPPD_NS_END
diff --git a/src/logger.h b/src/logger.h
index 7d3d7db..0446595 100644
--- a/src/logger.h
+++ b/src/logger.h
@@ -91,6 +91,9 @@ private:
 
     static int _max_pri;
 
+    // helpers for strerror_r
+    static std::string strerror_r_wrapper(int, char* s);
+    static std::string strerror_r_wrapper(char* s, char*);
 
 };
 
